eras/conway/impl/test/Test/Cardano/Ledger/Conway/ConwayEpochs.hs

```diff
@@ -0,0 +1,423 @@
+import Cardano.Ledger.Crypto (StandardCrypto)
+import Cardano.Ledger.DRepDistr (extractDRepDistr)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (..), coerceKeyRole)
+import Cardano.Ledger.Shelley.LedgerState (
+  IncrementalStake (..),
+  certVStateL,
+  curPParamsEpochStateL,
+  esLStateL,
+  lsCertStateL,
+  lsUTxOStateL,
+  nesELL,
+  nesEsL,
+  nesPdL,
+  prevPParamsEpochStateL,
+  totalObligation,
+  utxosDepositedL,
+  utxosGovStateL,
+  utxosStakeDistrL,
+  vsCommitteeStateL,
+  vsDRepsL,
+ )
+import Cardano.Ledger.TreeDiff (showExpr)
+import Cardano.Ledger.TxIn (TxId)
+import Cardano.Ledger.Val (Val (..))
+import Data.Default.Class (Default (..))
+import Data.Foldable (Foldable (..))
+import qualified Data.Map.Strict as Map
+import Data.Maybe (fromJust, fromMaybe)
+import Data.Maybe.Strict (StrictMaybe (..))
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Sequence.Strict as Seq
+import qualified Data.Text as T
+import Lens.Micro (to, (%~), (&), (.~), (^.))
+import Test.Cardano.Ledger.Common (
+  HasCallStack,
+  Spec,
+  describe,
+  shouldBe,
+  shouldSatisfy,
+ )
+import Test.Cardano.Ledger.Conway.ImpTest
+import Test.Cardano.Ledger.Core.KeyPair (mkAddr)
+import Test.Cardano.Ledger.Core.Rational (IsRatio (..))
+
+modifyPParams :: EraGov era => (PParams era -> PParams era) -> ImpTestM era ()
+modifyPParams f = do
+  modifyNES $ nesEsL . curPParamsEpochStateL %~ f
+  modifyNES $ nesEsL . prevPParamsEpochStateL %~ f
+
+registerDRep :: ImpTestM Conway (KeyHash 'DRepRole StandardCrypto)
+registerDRep = do
+  -- Register a DRep
+  khDRep <- freshKeyHash
+  _ <- submitBasicConwayTx "register DRep" $ \tx ->
+    tx
+      & bodyTxL . certsTxBodyL
+        .~ SSeq.singleton
+          ( mkRegDRepTxCert
+              (KeyHashObj khDRep)
+              zero
+              SNothing
+          )
+  do
+    dreps <- getsNES $ nesEsL . esLStateL . lsCertStateL . certVStateL . vsDRepsL
+    impIO $ dreps `shouldSatisfy` Map.member (KeyHashObj khDRep)
+  pure khDRep
+
+setupSingleDRep :: ImpTestM Conway (KeyHash 'DRepRole StandardCrypto)
+setupSingleDRep = do
+  khDRep <- registerDRep
+
+  khDelegator <- freshKeyHash
+  kpDelegator <- lookupKeyPair khDelegator
+  kpSpending <- lookupKeyPair =<< freshKeyHash
+  _ <- submitBasicConwayTx "Delegate to DRep" $ \tx ->
+    tx
+      & bodyTxL . outputsTxBodyL
+        .~ Seq.singleton
+          ( mkBasicTxOut
+              (mkAddr (kpSpending, kpDelegator))
+              (inject $ Coin 1000000)
+          )
+      & bodyTxL . certsTxBodyL
+        .~ Seq.fromList
+          [ ConwayTxCertDeleg $
+              ConwayRegDelegCert
+                (KeyHashObj khDelegator)
+                (DelegStakeVote (coerceKeyRole khDRep) (DRepCredential $ KeyHashObj khDRep))
+                zero
+          ]
+  pure khDRep
+
+submitProposal ::
+  GovAction Conway ->
+  ImpTestM Conway (GovActionId StandardCrypto)
+submitProposal constitutionAction = do
```

Comment:
Could this is named in a generic way like `govAction` instead of `constitutionAction` since it is not specific to being a `GovAction` of the `Constitution` variant alone, as per the body and type of the function?